src/main/java/org/openstreetmap/atlas/geography/atlas/raw/slicing/RawAtlasCountrySlicer.java

```diff
@@ -292,32 +726,319 @@ private boolean lineBelongsToSingleCountry(final Line line, final List<Geometry>
+        final List<RelationMember> nonClosedOuters = generateMemberList(relationIdentifier, outers,
+                false);
+
+        if (!nonClosedOuters.isEmpty())
+        {
+            final List<LinearRing> outerRings = buildRings(nonClosedOuters, relationIdentifier);
+            if (outerRings != null && !outerRings.isEmpty())
+            {
+                List<LinearRing> innerRings = null;
+                final MultiMap<Integer, Integer> outerToInnerMap = new MultiMap<>();
+                List<RelationMember> nonClosedInners = null;
+                if (inners != null && !inners.isEmpty())
+                {
+                    // Grab all non-closed inner ways
+                    nonClosedInners = generateMemberList(relationIdentifier, inners, false);
+                    innerRings = buildRings(nonClosedInners, relationIdentifier);
+                    if (innerRings != null && !innerRings.isEmpty())
+                    {
+                        // Try to find the outer ring that contains the inner ring
+                        findEnclosedRings(outerToInnerMap, outerRings, innerRings);
+                    }
+                }
+
+                for (int outerIndex = 0; outerIndex < outerRings.size(); outerIndex++)
+                {
+                    final LinearRing outerRing = outerRings.get(outerIndex);
+                    LinearRing[] holes = null;
+                    if (outerToInnerMap.containsKey(outerIndex))
+                    {
+                        final List<Integer> innerIndexes = outerToInnerMap.get(outerIndex);
+                        holes = new LinearRing[innerIndexes.size()];
+                        for (int innerIndex = 0; innerIndex < innerIndexes.size(); innerIndex++)
+                        {
+                            holes[innerIndex] = innerRings.get(innerIndexes.get(innerIndex));
+                        }
+                    }
+
+                    final com.vividsolutions.jts.geom.Polygon polygon = new com.vividsolutions.jts.geom.Polygon(
+                            outerRing, holes, JtsUtility.GEOMETRY_FACTORY);
+
+                    // Check if the polygon is valid. Sometimes polygons from relations can be
+                    // invalid, such as for large boundaries.
+                    if (!polygon.isValid())
+                    {
+                        logger.error("Polygon created by relation {} is invalid",
+                                relationIdentifier);
+                        return;
+                    }
+
+                    final List<LineString> borderLines;
+                    try
+                    {
+                        borderLines = this.countryBoundaryMap.clipBoundary(relationIdentifier,
+                                polygon);
+                    }
+                    catch (final Exception e)
+                    {
+                        logger.error("Error processing relation {}, message: {}, geometry: {}",
```

Comment:
Do we want to lose the cause of the exception here?

Reply:
Not at all, good catch!